test/e2e/resize_nodes.go

```diff
@@ -0,0 +1,331 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func resizeNodeInstanceGroup(size int) error {
+	// TODO: make this hit the compute API directly instread of shelling out to gcloud.
+	output, err := exec.Command("gcloud", "preview", "managed-instance-groups", "--zone="+testContext.CloudConfig.Zone,
+		"resize", testContext.CloudConfig.NodeInstanceGroup, fmt.Sprintf("--new-size=%v", size)).CombinedOutput()
+	if err != nil {
+		Logf("Failed to resize node instance group: %v", string(output))
+	}
+	return err
+}
+
+func waitForNodeInstanceGroupSize(size int) error {
+	for start := time.Now(); time.Since(start) < 4*time.Minute; time.Sleep(20 * time.Second) {
+		// TODO: make this hit the compute API directly instread of shelling out to gcloud.
+		output, err := exec.Command("gcloud", "preview", "managed-instance-groups", "--zone="+testContext.CloudConfig.Zone,
+			"describe", testContext.CloudConfig.NodeInstanceGroup).CombinedOutput()
+		if err != nil {
+			Logf("Failed to describe node instance group: %v", string(output))
+			continue
+		}
+		pattern := "currentSize: "
+		i := bytes.Index(output, []byte(pattern))
+		if i != -1 {
+			truncated := output[i+len(pattern):]
+			j := bytes.Index(truncated, []byte("\n"))
+			if j != -1 {
+				currentSize, err := strconv.Atoi(string(truncated[:j]))
+				if err == nil && currentSize == size {
+					Logf("Node instance group has reached the desired size %v", size)
+					return nil
+				}
+				if err == nil {
+					Logf("Waiting for node istance group size %v, current size %v", size, currentSize)
+					continue
+				} else {
+					Logf("Error while parsing current size: %v", err)
+				}
+			}
+		}
+		Logf("Waiting for node istance group size %v, current size unknown", size)
+	}
+	return fmt.Errorf("Timeout waiting for node instance group size to be %v", size)
+}
+
+func waitForClusterSize(c *client.Client, size int) error {
+	for start := time.Now(); time.Since(start) < 2*time.Minute; time.Sleep(20 * time.Second) {
+		nodes, err := c.Nodes().List(labels.Everything(), fields.Everything())
+		if err != nil {
+			Logf("Failed to list nodes: %v", err)
+			continue
+		}
+		if len(nodes.Items) == size {
+			Logf("Cluster has reached the desired size %v", size)
+			return nil
+		}
+		Logf("Waiting for cluster size %v, current size %v", size, len(nodes.Items))
+	}
+	return fmt.Errorf("Timeout waiting for cluster size to be %v", size)
+}
+
+func newService(name string) *api.Service {
```

Comment:
if there is anything specific to this test about the structs returned by `newService(...)`, `createService(...)`, or `newReplicationController(...)`, please rename them to something a bit more specific, as otherwise they have misleadingly generic names and are visible throughout the whole `e2e` package.